starknet_py/net/full_node_client.py

```diff
@@ -659,6 +657,110 @@ async def get_contract_nonce(
         res = cast(str, res)
         return int(res, 16)
 
+    # ------------------------------- Trace API -------------------------------
+
+    async def trace_transaction(
+        self,
+        tx_hash: Hash,
+    ) -> TransactionTrace:
+        """
+        For a given executed transaction, returns the trace of its execution, including internal calls.
+
+        :param tx_hash: Hash of the executed transaction.
+        :return: Trace of the transaction.
+        """
+        res = await self._client.call(
+            method_name="traceTransaction",
+            params={
+                "transaction_hash": tx_hash,
+            },
+        )
+        return cast(
+            TransactionTrace, TransactionTraceSchema().load(res, unknown=EXCLUDE)
+        )
+
+    async def simulate_transactions(
+        self,
+        transactions: List[AccountTransaction],
+        skip_validate: bool = False,
+        skip_fee_charge: bool = False,
+        block_hash: Optional[Union[Hash, Tag]] = None,
+        block_number: Optional[Union[int, Tag]] = None,
+    ) -> List[SimulatedTransaction]:
+        # pylint: disable=too-many-arguments
+        """
+        Simulates a given sequence of transactions on the requested state, and generates the execution traces.
+        If one of the transactions is reverted, raises CONTRACT_ERROR.
+
+        :param transactions: Transactions to be traced.
+        :param skip_validate: Flag checking whether there are enough funds in the account.
```

Comment:
Where do you have info on exact purpose of `skip_validate`? This is not a source of truth and might be a devnet specific thing, but [in starknet-jvm ](https://github.com/software-mansion/starknet-jvm/blob/f551c0827935aa79bda1b29898668a21f6af870f/lib/src/test/kotlin/starknet/account/StandardAccountTest.kt#L777) we actually test this flag for transactions with removed signature.